feat(regex-parser): parse simple `TemplateLiterals`

Sysix opened the pull request:
This PR adds the function to parse `TemplateLiteral`s.
If the `new RegExp(`${variable}`)` or `new RegExp("", `${foo}`)` the parsing will be skipped, this is a different behavior then for `StringLiteral`s.
https://github.com/eslint/eslint/blob/5082fc206de6946d9d4c20e57301f78839b3b9f2/tests/lib/rules/no-misleading-character-class.js#L91-L93
I would like to have a strict algo, when a regexp will be parsed and not. Currently, every rule does it by its own. 
https://github.com/oxc-project/oxc/blob/2cd4c7b42cd7d4d455dd1f64ee453e05d5da26ef/crates/oxc_linter/src/rules/eslint/no_control_regex.rs#L322

Moved some structs into the parent directory, so it can be consumed by both parsers. I guess `ParserTrait` could be helpful to share common methods like
-  consume_hex_digits
- parse_unicode_escape_sequence

Sysix commented:
* **#13265** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13265?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13265).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_regular_expression/src/parser/reader/string_literal_parser/mod.rs:
```diff
@@ -240,4 +243,14 @@ without continuation""#,
+
+        let result = Parser::new(source_text, Options::default()).parse();
+
+        assert!(result.is_ok(), "Expected to parse: {source_text} but failed: {:?}", result.err());
```
@leaysgur I do not have deep knowledge about the regex parser. 

Line breaks in template literals are allowed with regex. I want to cover the test case from `no-misleading-character-class`:
https://github.com/eslint/eslint/blob/0c9999c2a682151cf13bb3a4f8916930678c2f9b/tests/lib/rules/no-misleading-character-class.js#L841-L858

Can you guide me into the right direction or contribute to the PR :)?
PS: The naming is now for `StringLiteral(Kind)` is now wrong. Happy for name suggestions^^"

camc314 commented:
@Sysix somehow this has ended up with conflicts 🫠

leaysgur commented:
Sorry for the slow response! 🙇🏻

I try to review by Saturday(during the flight), or at the latest by next Tuesday.

Sysix commented:
No need to rush, glad you're offering your help with the regex parser

leaysgur commented on crates/oxc_regular_expression/src/parser/reader/reader_impl.rs:
```diff
@@ -45,7 +63,7 @@ impl<'a> Reader<'a> {
             index: 0,
             // If `parse_string_literal` is `true`, the first character is the opening quote.
```
```suggestion
            // If `parse_string_or_template_literal` is `true`, the first character is the opening quote.
```

leaysgur commented on crates/oxc_regular_expression/src/parser/reader/template_literal_parser/README.md:
```diff
@@ -0,0 +1,5 @@
+
+Implements ECMAScript® 2025 Language Specification
+
+- https://tc39.es/ecma262/2025/multipage/ecmascript-language-lexical-grammar.html#sec-template-literal-lexical-components
```
Should note about we only support `NoSubstitutionTemplate` for now?

leaysgur commented on crates/oxc_regular_expression/src/parser/reader/template_literal_parser/parser_impl.rs:
```diff
@@ -0,0 +1,573 @@
+}
+
+impl Parser {
+    // This is public because it is used in `parse_regexp_literal()`.
+    pub fn handle_code_point(
```
```suggestion
    fn handle_code_point(
```

leaysgur commented on crates/oxc_regular_expression/src/parser/reader/template_literal_parser/parser_impl.rs:
```diff
@@ -0,0 +1,573 @@
+        if self.peek() == Some(CR) && self.peek2() == Some(LF) {
+            self.advance();
+            self.advance();
+            // NOTE: CR+LF can not represent as a single code point.
+            // I don't know the best way to handle this.
+            // To distinguish this from CR and LF, structural change is needed...
```
```suggestion
```

No need to repeat? 👀

leaysgur commented on crates/oxc_regular_expression/src/parser/reader/template_literal_parser/parser_impl.rs:
````diff
@@ -0,0 +1,573 @@
+    //   <LS>
+    //   <PS>
+    // ```
+    fn parse_line_terminator(&mut self) -> Option<u32> {
````
You can use `character::is_line_terminator` and inline it like https://github.com/oxc-project/oxc/blob/ca4f52fa62af96b72aec0c9ab8975ceb9d5b84bb/crates/oxc_regular_expression/src/parser/reader/string_literal_parser/parser_impl.rs#L153

leaysgur replied:
OK, my question was about usage at L156.

How about writing it like this?

```rs
if ch == '\\' || ch == '`' || is_line_terminator(ch) {
```

`self.parse_line_terminator().is_some()` should not call `advance()` for checking `SourceCharacter but not one of ` or \ or $ or LineTerminator`?

leaysgur reviewed (changes requested):
@Sysix Sorry for the delay!

As you noticed, there is still room for reducing duplicate lines, but I think we can address that at another time.

leaysgur reviewed (approved):
🚀

Sysix committed:
feat(regex-parser): parse simple `TemplateLiterals` (#13265)

This PR adds the function to parse `TemplateLiteral`s.
If the `new RegExp(`${variable}`)` or `new RegExp("", `${foo}`)` the parsing will be skipped, this is a different behavior then for `StringLiteral`s.
https://github.com/eslint/eslint/blob/5082fc206de6946d9d4c20e57301f78839b3b9f2/tests/lib/rules/no-misleading-character-class.js#L91-L93
I would like to have a strict algo, when a regexp will be parsed and not. Currently, every rule does it by its own.
https://github.com/oxc-project/oxc/blob/2cd4c7b42cd7d4d455dd1f64ee453e05d5da26ef/crates/oxc_linter/src/rules/eslint/no_control_regex.rs#L322

Moved some structs into the parent directory, so it can be consumed by both parsers. I guess `ParserTrait` could be helpful to share common methods like
-  consume_hex_digits
- parse_unicode_escape_sequence
